File: bindings/java/src/main/java/org/indunet/robot/bus/ServiceHandle.java
package org.indunet.robot.bus;

import com.sun.jna.Pointer;

/** Opaque service server handle returned by {@link Node#createService}. */
public final class ServiceHandle implements AutoCloseable {
    private final Pointer node;
    Pointer ptr;
    private boolean destroyed;

    ServiceHandle(Pointer node, Pointer ptr) {
        this.node = node;
        this.ptr = ptr;
    }

    public String serviceName() {
        if (ptr == null) {
            return "";
        }
        Pointer s = RobotBusC.Holder.INSTANCE.robot_bus_service_handle_name(ptr);
        return NativeUtils.takeCString(s);
    }

    public void destroy() {
        if (destroyed || ptr == null) {
            return;
        }
        Errors.check(
                RobotBusC.Holder.INSTANCE.robot_bus_node_destroy_service(node, ptr),
                "destroy_service");
        destroyed = true;
    }

    @Override
    public void close() {
        destroy();
        if (ptr != null) {
            RobotBusC.Holder.INSTANCE.robot_bus_service_handle_free(ptr);
            ptr = null;
        }
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/ServiceHandle.kt
package org.indunet.robot.bus

import com.sun.jna.Pointer

/** Opaque service server handle returned by [Node.createService]. */
class ServiceHandle
internal constructor(
    private val node: Pointer?,
    internal var ptr: Pointer?,
) : AutoCloseable {
    private var destroyed = false

    fun serviceName(): String {
        val p = ptr ?: return ""
        return NativeUtils.takeCString(RobotBusC.Holder.INSTANCE.robot_bus_service_handle_name(p))
    }

    fun destroy() {
        if (destroyed || ptr == null) return
        Errors.check(
            RobotBusC.Holder.INSTANCE.robot_bus_node_destroy_service(node, ptr),
            "destroy_service",
        )
        destroyed = true
    }

    override fun close() {
        destroy()
        ptr?.let { RobotBusC.Holder.INSTANCE.robot_bus_service_handle_free(it) }
        ptr = null
    }
}
